### PR TITLE
Move to optional canonical id on identifiable object

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

Resolving having identifiers for both items and works.

### Who is this change for?

Folks who want identified items in the API.

### Have the following been considered/are they needed?

- [ ] Deployed new versions

### CODE REVIEW

**Comment by kenoir** [comment]
Depends https://github.com/wellcometrust/platform-api/pull/829

**Comment by kenoir** [comment]
For https://github.com/wellcometrust/platform-api/issues/572 & https://github.com/wellcometrust/platform-api/issues/573

**Comment by kenoir** [comment]
This PR will not include the changes to add `Items` as this is already a significant data model update. 

We will need to reindex into the new data model before populating `Items`.

File: api/docker-compose.yml [hunk 1]
```diff
@@ -7,3 +7,4 @@ elasticsearch:
     - "http.host=0.0.0.0"
     - "transport.host=0.0.0.0"
     - "cluster.name=wellcome"
+    - "logger.level=DEBUG"
```
Comment (alexwlchan) [inline api/docker-compose.yml]: Do we want to check in this change?

File: api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala [hunk 2]
```diff
@@ -73,28 +73,28 @@ case object DisplayWork {
...
+      JsonUtil.fromJson[Work](document).get
 
     DisplayWork(
-      id = identifiedWork.canonicalId,
-      title = identifiedWork.work.title,
-      description = identifiedWork.work.description,
-      lettering = identifiedWork.work.lettering,
-      createdDate = identifiedWork.work.createdDate,
+      id = work.id,
+      title = work.title,
```
Comment (alexwlchan) [inline api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala]: Not having this layer of indirection is very nice.

File: api/src/test/scala/uk/ac/wellcome/platform/api/ApiWorksTest.scala [hunk 3]
```diff
@@ -116,16 +118,16 @@ class ApiWorksTest
   }
 
   it("should return a single work when requested with id") {
-    val identifiedWork =
-      identifiedWorkWith(
+    val work =
+      workWith(
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/ApiWorksTest.scala]: nit: can probably pull this up onto the previous line

File: api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala [hunk 4]
```diff
@@ -14,19 +14,19 @@ trait WorksUtil {
...
   val agent = Agent("a person")
 
-  def convertWorkToDisplayWork(work: IdentifiedWork) = DisplayWork(
-    work.canonicalId,
-    work.work.title,
-    work.work.description,
-    work.work.lettering,
-    work.work.createdDate,
-    work.work.creators
+  def convertWorkToDisplayWork(work: Work) = DisplayWork(
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala]: Possibly one to defer to a separate patch, but maybe we should just have an `apply()` method on `DisplayWork` that takes an instance of `Work` as an argument? It’s what we have on several of the other Display classes.

File: api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala [hunk 5]
```diff
@@ -35,40 +35,40 @@ trait WorksUtil {
...
 
-  def identifiedWorkWith(canonicalId: String, title: String): IdentifiedWork =
-    IdentifiedWork(canonicalId,
-                   Work(identifiers =
+  def workWith(canonicalId: String, title: String): Work =
+    Work(Some(canonicalId),
+                   identifiers =
                           List(SourceIdentifier(IdentifierSchemes.miroImageNumber, "5678")),
-                        title = title))
+                        title = title)
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala]: this indentation is all over the place

File: api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala [hunk 6]
```diff
@@ -35,40 +35,40 @@ trait WorksUtil {
...
+  def workWith(canonicalId: String,
                          title: String,
                          description: String,
                          lettering: String,
                          createdDate: Period,
-                         creator: Agent): IdentifiedWork = IdentifiedWork(
-    canonicalId = canonicalId,
-    work = Work(
+                         creator: Agent): Work = Work(
+    canonicalId = Some(canonicalId),
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala]: more indentation that's quite messy

File: api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala [hunk 7]
```diff
@@ -45,16 +45,19 @@ class DisplayItemTest extends FunSpec with Matchers {
 
   it("should read an Item as a DisplayItem correctly") {
 
-    val internalItem = Item(
-      identifiers = List(identifier),
-      locations = List(location)
-    )
+    val item = Item(
+        Some("foo"),
```
Comment (alexwlchan) [inline api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala]: It would be nice if this was either a keyword argument (`Item(id = Some("foo") ...`) or the `"foo"` was more clearly an identifier. Or both.

File: common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package uk.ac.wellcome.models
+
+case class UnidentifiableException()
+    extends Exception("Can't find canonicalId for this Identifiable")
+
+trait Identifiable {
+  val canonicalId: Option[String]
+  def id: String = canonicalId.getOrElse(
+    throw UnidentifiableException()
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala]: Worth a test that we really do throw this exception if you try to index

* A work without an ID?
* A work with an ID with an item without?

(Or have I missed those tests?)

File: api/docker-compose.yml (hunk 1 above)
Comment (kenoir) [inline api/docker-compose.yml]: I think so. Trying to debug what's going on when a test fails is pretty much useless without it.

File: api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala (hunk 5 above)
Comment (kenoir) [inline api/src/test/scala/uk/ac/wellcome/platform/api/WorksUtil.scala]: take it up with `scalafmt`

File: api/docker-compose.yml (hunk 1 above)
Comment (alexwlchan) [inline api/docker-compose.yml]: 👍

Comment (alexwlchan) [inline api/docker-compose.yml]: (I’d missed this was only on the tests.)

File: common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala (hunk 8 above)
Comment (kenoir) [inline common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala]: 👍

File: api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala [hunk 9]
```diff
@@ -73,30 +99,9 @@ case object DisplayWork {
   }
 
   private def jsonToDisplayWork(document: String, includes: WorksIncludes) = {
-    val identifiedWork =
-      JsonUtil.fromJson[IdentifiedWork](document).get
+    val work =
+      JsonUtil.fromJson[Work](document).get
```
Comment (alexwlchan) [inline api/src/main/scala/uk/ac/wellcome/platform/api/models/DisplayWork.scala]: nit: pull this onto a single line, lose the newline?

File: common/src/main/scala/uk/ac/wellcome/models/SourceIdentifier.scala [hunk 10]
```diff
@@ -0,0 +1,4 @@
+package uk.ac.wellcome.models
+
+/** An identifier received from one of the original sources */
+case class SourceIdentifier(identifierScheme: String, value: String)
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/models/SourceIdentifier.scala]: 👌

File: common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    val result = JsonUtil.fromJson[Item](identifiedItemJson)
+
+    result.isSuccess shouldBe true
+    result.get shouldBe identifiedItem
+  }
+
+  it("should throw an UnidentifiableException when trying to get the id of an unidentified Item") {
+    an [UnidentifiableException] should be thrownBy unidentifiedItem.id
+  }
+}
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala]: This test looks good, but most of it arguably belongs in a separate patch?

👍 for the UnidentifiableException test

File: common/src/test/scala/uk/ac/wellcome/models/WorkTest.scala [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    val result = JsonUtil.fromJson[Work](identifiedWorkJson)
+
+    result.isSuccess shouldBe true
+    result.get shouldBe identifiedWork
+  }
+
+  it("should throw an UnidentifiableException when trying to get the id of an unidentified Work") {
+    an [UnidentifiableException] should be thrownBy unidentifiedWork.id
+  }
+}
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/models/WorkTest.scala]: Ditto above

File: ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexerTest.scala [hunk 13]
```diff
@@ -25,32 +25,32 @@ class IdentifiedWorkIndexerTest
...
+                      title = title))
       .get
   }
 
   it("should insert an identified unified item into Elasticsearch") {
 
-    val identifiedWorkString =
-      identifiedWorkJson("5678", "1234", "An identified igloo")
+    val workString =
+      workJson("5678", "1234", "An identified igloo")
```
Comment (alexwlchan) [inline ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexerTest.scala]: nit: move onto a single line?

File: ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexerTest.scala [hunk 14]
```diff
@@ -25,32 +25,32 @@ class IdentifiedWorkIndexerTest
...
       .toJson(
-        IdentifiedWork(
-          canonicalId = canonicalId,
-          work = Work(identifiers =
+        Work(
+            canonicalId = Some(canonicalId),
+                identifiers =
                         List(SourceIdentifier(IdentifierSchemes.miroImageNumber, sourceId)),
-                      title = title)))
+                      title = title))
```
Comment (alexwlchan) [inline ingestor/src/test/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexerTest.scala]: wtf indentation
